ASoC: SOF: (Intel HDA) Add support for DSPless debug mode

ujfalusi opened the pull request:
Replacing #3958, now with the correct branch!

Hi,

this series will enable the use of the SOF Linux stack without DSP offloading.

In this mode no firmware loading will happen, the topology parsing is going to be reduced to only look for the DAI widget(s) and the IPC dependent callbacks are going to be ignored.

This mode can give another level of hardware verification on platforms where the DSP can be ignored and the audio interfaces can be tested directly.

On Intel platforms we can use this mode to verify programming flows against the legacy HDA stack if there is a need to debug in that level and don't have the DSP programming sequences interfering.

The use of DSPless mode is governed by the SOF_DBG_DSPLESS_MODE flag which is only going to be set if the user sets sof_debg=0x8000 and the platform advertises that the DSPless mode is supported on them.

ujfalusi commented:
Changes since v1:
- Use sdev->dspless_mode_selected runtime for decisions instead of the sof_debug flag
- all minor comments addressed with the exception of the use of a temporary tplg pointer

ujfalusi commented:
Changes since v2:
- set constraint on the supported formats for HDA in DSPless mode: Only S16 and S32 formats are supported (S24_LE is not supported).

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -718,6 +796,18 @@ void hda_set_dai_drv_ops(struct snd_sof_dev *sdev, struct snd_sof_dsp_ops *ops)
 	int i;
 
+	if (sdev->dspless_mode_selected) {
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
```
didn't I change this in my HDaudio cleanups? I don't have the code handy but I thought I changed that to CONFIG_SND_SOC_SOF_HDAUDIO_CODEC

ujfalusi replied:
oh, yes you did, but it still works :o

ujfalusi commented:
Changes since v3:
- do not allow DSPless mode to be enabled if the NOCODEC mode is selected
- Do not connect non HDA DAIs to allow mixed topology loading but make sure that the PCMs with non HDA DAIs are not usable by user.
- Use the correct CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC in hda-dai.c

plbossart commented:
> * do not allow DSPless mode to be enabled if the NOCODEC mode is selected

This needs to be platform-dependent @ujfalusi. If we hard-code this we are going to un-hard-code it soon.

ujfalusi commented:
Changes since v4:
- Lift additional restriction in core on the DSPless mode (NOCODEC and INTEL_HDA dai)
- Introduce `hda_get_interface_mask()` in intel/hda.c to get an interface mask about the currently supported interfaces
 - the mask is Intel audio IP _and_ DSPless mode dependent.
 - Only do discovery of I2S, DMIC and SDW if they are supported.

The only issue with the implementation is that if the DSPless mode is selected and also the NOCODEC is forced we will end up with a card which is not usable as nocodec topologies usually falling back to use SSP and DMIC and they are not usable without DSP.

Anyways, this PR works OK on UP2 when using the pcm512x ACPI patching and the resulting card will fall back to a generic HDA card:
```
[12958.191985] sof-audio-pci-intel-apl 0000:00:0e.0: Switching to DSPless mode
...
[12958.249559] snd_sof:sof_machine_register: sof-audio-pci-intel-apl 0000:00:0e.0: created machine skl_hda_dsp_generic
[12958.249580] snd_sof_pci:sof_pci_probe_complete: sof-audio-pci-intel-apl 0000:00:0e.0: Completing SOF PCI probe
[12958.267497] snd_sof:snd_sof_load_topology: sof-audio-pci-intel-apl 0000:00:0e.0: loading topology:intel/sof-tplg/sof-hda-generic-idisp.tplg
```
and
```
# aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: sofhdadsp [sof-hda-dsp], device 1: HDMI1 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofhdadsp [sof-hda-dsp], device 2: HDMI2 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofhdadsp [sof-hda-dsp], device 3: HDMI3 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

ranj063 commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -603,6 +603,84 @@ static const struct snd_soc_dai_ops ipc4_hda_dai_ops = {
+	hstream->format_val = snd_hdac_calc_stream_format(params_rate(params),
+							  params_channels(params),
+							  params_format(params),
+							  sig_bits, 0);
```
is saving enough? Dont we need to call snd_hdac_ext_stream_setup() with the format?

ujfalusi replied:
afaik when the HDA-DMA is coupled the format only need to be set in one place and it is going to propagate to the link side

ranj063 replied:
then what does this saving really do? is it needed at all?

plbossart replied:
this is indeed a problem, because we set this format_val information in hda-pcm.c based on the host hw_params only.
````c
hda_dsp_pcm_hw_params	
     hstream->format_val = rate | bits | (params_channels(params) - 1);


hda_dsp_stream_hw_params
     	/* program stream format */
	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR,
				sd_offset +
				SOF_HDA_ADSP_REG_SD_FORMAT,
				0xffff, hstream->format_val);
````

this means we never take into account the codec_dai information. It's possible that things work by accident but it doesn't seem quite right.

ujfalusi replied:
We need to save this for the FE side to use.
With DSP in picture the BE (Link) side and FE (host) side of the HDA-DMA can have different format configured and the conversion is done inside of the DSP, but when we don't have the DSP we need the codec side information used on the host side so the format is correct.

plbossart replied:
Completely agree we need to keep this information, my point is that we shouldn't program the format twice, first with a value that ignores the codec settings and second with a value that takes codec settings into account.

I don't even know which of the two is used today, the order between FE and BE keeps changing so it's anyone's guess what is used.

Maybe we should make the first assignment in hda-pcm.c conditional on the presence of the DSP?

ujfalusi replied:
Like in the first chunk change in the hda-pcm.c in this patch? At line 114

plbossart replied:
ah yes, that will do @ujfalusi. thanks for the pointer. We can close this one.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -392,15 +393,21 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
+	     !sof_ops(sdev)->block_write || !sof_ops(sdev)->send_msg ||
+	     !sof_ops(sdev)->load_firmware || !sof_ops(sdev)->ipc_msg_data)) {
+		sof_ops_free(sdev);
+		dev_err(dev, "error: missing mandatory DSP ops\n");
```
remove error: prefix in all this code?

ujfalusi replied:
oops, true

plbossart reviewed (commented):
@ujfalusi we're going to need some more time to review the last patches, so can we push the initial 13-odd patches in a separate PR to reduce the volume of patches to carry in this draft?

more comments below.

ujfalusi commented:
> @ujfalusi we're going to need some more time to review the last patches, so can we push the initial 13-odd patches in a separate PR to reduce the volume of patches to carry in this draft?

I have separated the ops optionality patches s they are one topic (but in itself it does not have much reason to exist): https://github.com/thesofproject/linux/pull/3986

plbossart commented:
@ujfalusi can you rebase so that we look at the latest and greatest code on this branch?

ujfalusi commented:
Changes since v5:
- rebased
- Separated the skip unsupported interfaces from the DSPless mode patch to make it easier to focus on the real changes.

plbossart commented:
@ujfalusi would you be able to rebase this PR on top of #/3972? It'd be good if @ranj063 and I can take a look tomorrow morning (Nov 18) before the week-end and the Turkey break. Thanks!

plbossart commented:
https://github.com/thesofproject/linux/pull/3972 is merged, please rebase @ujfalusi

ujfalusi commented:
Changes since v6:
- reworked on top of sof-dev with #3972 merged
 - Things might need to be revisited since #3972 made IPC mode semi mandatory for the HDA code.
 - If there is no swidget allocated, it implies that no IPC is used, which is the case of the DSPless mode

ujfalusi commented:
Oops, something is not quite how it should be in DSP mode...

ujfalusi commented:
Changes since v7:
- DSP mode fixed (broke due to returning uninitialized 'ops' from `hda_dai_get_ops()` on the second call)
- dropped the 'error:' prefix from a print in core

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -192,6 +192,9 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 
+	target_state = snd_sof_dsp_power_target(sdev);
+	pm_state.event = target_state;
+
```
@ranj063 has a different solution for the same problem it seems @ujfalusi 

see https://github.com/thesofproject/linux/pull/4019/commits/297cb2f58f263c9b5fa86dbf605ebcfdb487e14e

which one should we take?

ujfalusi replied:
They are identical in function, we can take @ranj063's version if that makes in earlier (I'm sure it will)

ranj063 replied:
this patch needs to go now that my PR is merged right @ujfalusi ?

plbossart replied:
indeed this seems to be the source of conflicts?

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -310,7 +310,7 @@ sof_unprepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widg
 
 	/* return if the widget is in use or if it is already unprepared */
-	if (!swidget->prepared || swidget->use_count > 1)
+	if (!swidget || !swidget->prepared || swidget->use_count > 1)
```
and this one is being changed by @bardliao in https://github.com/thesofproject/linux/pull/4022

ujfalusi replied:
Probably they are aiming for the same issue. Will be dropped, but it is needed atm.

plbossart replied:
drop this patch?

ujfalusi replied:
We still need to have the !swidget check

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -310,6 +310,27 @@ static const struct hda_dai_widget_dma_ops hda_ipc3_dma_ops = {
+					  struct hdac_ext_stream *hext_stream,
+					  unsigned int format_val)
+{
+	hext_stream->hstream.format_val = format_val;
```
maybe add a comment that we store the format_val for later use? Otherwise this looks useless.

Actually looking at this twice it does look useless.

ujfalusi replied:
It is not useless. The maxbps (sig_bits) of the real codec is only know in the BE. In DSPless mode we need to use this information on the BE, but on there it is not available (we have dummy-codec).
It needs to be saved, otherwise the DMA will be miss configured.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -56,13 +63,22 @@ hda_dai_get_ops(struct snd_pcm_substream *substream, struct snd_soc_dai *cpu_dai
-		const struct hda_dai_widget_dma_ops *ops =
-			hda_select_dai_widget_ops(sdev, swidget);
-
+		ops = hda_select_dai_widget_ops(sdev, swidget);
```
I was not able to follow the change from the initial code written by @ranj063 

In one case sdai->platform_private is required and not in the other.

ranj063 replied:
there's no swidget and so there's no sdai with dspless mode @plbossart. so there's cant be a platform_private either

ujfalusi replied:
This change might not be needed if the swidget allocation way works.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -56,13 +63,22 @@ hda_dai_get_ops(struct snd_pcm_substream *substream, struct snd_soc_dai *cpu_dai
+
+	if (!swidget) {
+		ops = hda_select_dai_widget_ops(sdev, NULL);
+		if (!ops || !ops->get_hext_stream || !ops->setup_hext_stream)
```
why is setup_hext_stream mandatory now?

ranj063 replied:
I don't think we can have different set of mandatory ops for different instances of the ops. setup() is optional for this exact reason.

ujfalusi replied:
This is exactly what I was saying for the Abstraction PR. This two op is mandatory for normal mode and for DSPless.
In fact, the hext_stream ops are mandatory for the normal mode.
I can drop the mandatory check, but it is mandatory.

And the fact is that different modes (normal chain and DSPless) very well needs different sets of mandatory callbacks.

ranj063 replied:
@ujfalusi different modes needing different ops doesnt make the ops mandatory. Its a bit like saying the IPC ops we have for IPC3/IPC4 will need to have different set of mandatory ops right? But if you think setup is mandatory for all cases, lets make it mandatory

plbossart commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -486,7 +486,7 @@ int hda_dsp_stream_hw_params(struct snd_sof_dev *sdev,
 	int ret;
 	u32 dma_start = SOF_HDA_SD_CTL_DMA_START;
-	u32 mask;
+	u32 mask = BIT(hstream->index);
```
actually the initial code is rotten already. We assign hstream by dereferencing hext_stream BEFORE testing hext_stream. This is not right. both hstream and mask assignments need to be placed after the test for if (!hext_stream)

ujfalusi replied:
Hah, that's true. I'll fix that as well with a separate patch.

plbossart commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -672,20 +670,23 @@ int hda_dsp_stream_hw_free(struct snd_sof_dev *sdev,
-	spin_unlock_irq(&bus->reg_lock);
+	if (!sdev->dspless_mode_selected) {
+		struct hdac_bus *bus = sof_to_bus(sdev);
+		u32 mask = 0x1 << hstream->index;
```
BIT(hsteam->index)  for consistency?

ujfalusi replied:
Sure

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -47,32 +47,34 @@
-				 BIT(SOF_DAI_INTEL_HDA) | BIT(SOF_DAI_INTEL_ALH);
+		interface_mask[0] = BIT(SOF_DAI_INTEL_SSP) | BIT(SOF_DAI_INTEL_DMIC) |
+				    BIT(SOF_DAI_INTEL_HDA) | BIT(SOF_DAI_INTEL_ALH);
+		interface_mask[1] = BIT(SOF_DAI_INTEL_HDA);
```
it's probably a different patch, this is just a table change, maybe squash this with the initial change?

ujfalusi replied:
No, it can not be done that way. In there the array makes no sense as we did not have dspless mode.
And we will need another flag for this since the same rule applies to the chainDMA (can only be used with HDA)

plbossart commented on sound/soc/sof/intel/pci-apl.c:
```diff
@@ -29,6 +29,7 @@ static const struct sof_dev_desc bxt_desc = {
 	.chip_info = &apl_chip_info,
 	.ipc_supported_mask	= BIT(SOF_IPC) | BIT(SOF_INTEL_IPC4),
 	.ipc_default		= SOF_IPC,
+	.dspless_mode_supported	= true,		/* Only supported for HDaudio */
```
this could also be set with the test in the interface version. There's no good reason why we set the interfaces with a function and this flag as a constant

ujfalusi replied:
Yeah, but I have not been able to get my hands on on other than TGL and APL class device. It should work on others, but I can not say for sure...
But then we will have issue in core to block the dspless mode on platforms that it is _known_ to not work (AMD, iMX, MTK, Atom).
We should have this flag.

plbossart replied:
ok but then let's add it everywhere, not just APL and TGL

ujfalusi replied:
Hrm, HDA devices should work just fine in DSPless mode. I guess I can flag them even if I was not able to run the DSPless mode on them?

plbossart reviewed (commented):
Couple of comments @ujfalusi , I wasn't able to follow some of the changes on why we need a setup function and other options. Nitpicks otherwise.

ranj063 commented on sound/soc/sof/core.c:
```diff
@@ -365,6 +365,20 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
 
+	if (sof_debug_check_flag(SOF_DBG_DSPLESS_MODE)) {
+		if (plat_data->desc->dspless_mode_supported) {
+			if (sof_debug_check_flag(SOF_DBG_FORCE_NOCODEC)) {
```
can we combine these 2 inot 1 condition with && and the else can have a single info saying "DSPless mode sectiond ignored"? That will reduce the indendation a bit

ujfalusi replied:
It is a good information that the DSPless mode is not selected due to platform not supporting it and the user request is ignored.

ujfalusi replied:
it is not the same as the user has not asked for it in the first place.

ranj063 commented on sound/soc/sof/sof-priv.h:
```diff
@@ -516,6 +517,16 @@ struct snd_sof_dev {
+	/*
+	 * When true the DSP is not used.
+	 * It is set under the following condition:
+	 * User set the SOF_DBG_DSPLESS_MODE flag in sof_debug module parameter
```
typo sets

ujfalusi replied:
OK

ranj063 commented on sound/soc/sof/core.c:
```diff
@@ -367,13 +373,8 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
-				dev_info(dev, "Switching to DSPless mode\n");
-				sdev->dspless_mode_selected = true;
-			}
+			dev_info(dev, "Switching to DSPless mode\n");
```
why did you add the check for NOCODEC only to remove it in the very next patch?

ujfalusi replied:
Oh, Right, I will drop it from the first patch.

ranj063 commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -56,13 +63,22 @@ hda_dai_get_ops(struct snd_pcm_substream *substream, struct snd_soc_dai *cpu_dai
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(cpu_dai->component);
 	struct snd_sof_widget *swidget = w->dobj.private;
-	struct snd_sof_dai *sdai = swidget->private;
+	const struct hda_dai_widget_dma_ops *ops;
```
@ujfalusi I think something along these lines as below might look better. What do you think?
```
static const struct hda_dai_widget_dma_ops *
hda_dai_get_ops(struct snd_pcm_substream *substream, struct snd_soc_dai *cpu_dai)
{
	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(cpu_dai, substream->stream);
	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(cpu_dai->component);
	struct snd_sof_widget *swidget = w->dobj.private;
	const struct hda_dai_widget_dma_ops *ops;
	struct snd_sof_dai *sdai;

	/* return already set ops */
	if (swidget) {
		sdai = swidget->private;
		if (sdai->platform_private)
			return sdai->platform_private;
	}

	/* select the op */
	ops = hda_select_dai_widget_ops(sdev, swidget);
	if (!ops)
		return NULL;

	/* check if mandatory ops are set */
	if (!ops || !ops->get_hext_stream)
		return NULL;

	/* set the ops for DSP mode */
	if (swidget)
		sdai->platform_private = ops;
	
	return ops;
}
```

ranj063 replied:
@ujfalusi I also have an alternate suggestion. We have a the snd_soc_dapm_widget fr the BE DAI, what we're missing is swidget. Now, if we could modify the topology ops like this, you wont have to touch this function or hda_dai_config() for that matter

```
static int sof_dspless_widget_ready(struct snd_soc_component *scomp, int index,
			    struct snd_soc_dapm_widget *w,
			    struct snd_soc_tplg_dapm_widget *tw)
{
	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);

	swidget = kzalloc(sizeof(*swidget), GFP_KERNEL);
	if (!swidget)
		return -ENOMEM;

	switch (w->id) {
	case snd_soc_dapm_dai_in:
	case snd_soc_dapm_dai_out:
		struct snd_sof_dai *dai;

		dai = kzalloc(sizeof(*dai), GFP_KERNEL);
                if(!dai)
                        return -ENOMEM;
		ret = sof_connect_dai_widget(scomp, w, tw, dai);
		if (ret < 0) {
			kfree(swidget);
                        kfree(dai);
			return ret;
		}
               swidget->private = dai;
               break;
	default:
		break;
	}

	w->dobj.private = swidget;
	return 0
}

static int sof_dspless_widget_unload(struct snd_soc_component *scomp,
			     struct snd_soc_dobj *dobj)
{

	struct snd_sof_widget *swidget = dobj->private;

	if (!swidget)
		return 0;

        kfree(swidget->private);
	kfree(swidget);
}
```

ujfalusi replied:
I have thought about the second way to allocate the swidget, let me test on Monday if it brings other issues or not.

ujfalusi replied:
Right, so this is not going to work without doing a full blown topology parsing in case of DSPless mode as we need to set up all the things to get the information (`ipc4_copier->dai_type` or `dai_config->type`)
I think I will replace the `!swidget` checks with explicit `dspless_mode_selected` queries to not obscure the code.
Yep, less generic, but easier to understand.

ujfalusi commented:
Changes since v8:
- separate patch to correct the hda-stream use before NULL check of hstream
- use dspless_mode_selected flag instead of !swidget check where it is possible to make the code easier to follow
- minor comments  addressed

ranj063 commented on sound/soc/sof/core.c:
```diff
@@ -365,6 +365,15 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
+			dev_info(dev, "Switching to DSPless mode\n");
+			sdev->dspless_mode_selected = true;
+		} else {
+			dev_info(dev, "DSPless mode is not supported by the platform\n");
```
can this be downgraded to a dev_dbg?

plbossart replied:
I personally prefer the dev_info, this would tell us right away if a user mucks with the kernel parameters or if the CI setup is incorrect. I would even make this one a dev_err().

ujfalusi replied:
dev_info() is I think the right level. I would reserve dev_err for real errors... Can be dev_warn() if we want to make more noise...

ranj063 commented on sound/soc/sof/intel/hda.c:
```diff
@@ -44,6 +44,37 @@
 #define EXCEPT_MAX_HDR_SIZE	0x400
 #define HDA_EXT_ROM_STATUS_SIZE 8
 
+static u32 hda_get_interface_mask(struct snd_sof_dev *sdev)
```
typo in commit message: Not all interfaces "are" available...no point "in executing"....

ujfalusi replied:
OK

ranj063 commented:
and oh time to undraft this PR I think

plbossart commented:
@ujfalusi can you please rebase and solve conflicts?

plbossart commented on sound/soc/sof/intel/pci-tgl.c:
```diff
@@ -247,6 +253,7 @@ static const struct sof_dev_desc rpl_desc = {
 	.chip_info = &tgl_chip_info,
 	.ipc_supported_mask	= BIT(SOF_IPC) | BIT(SOF_INTEL_IPC4),
 	.ipc_default		= SOF_IPC,
+	.dspless_mode_supported	= true,		/* Only supported for HDaudio */
```
any reason why we only add this mode for APL and TGL? CNL/ICL/MTL are missing.

ujfalusi replied:
I only tested on APL and TGL, but I will allow DSPless mode for the other platforms having HDA.

plbossart commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -482,9 +482,8 @@ int hda_dsp_stream_hw_params(struct snd_sof_dev *sdev,
-	int sd_offset = SOF_STREAM_SD_OFFSET(hstream);
-	int ret;
+	struct hdac_stream *hstream;
+	int sd_offset, ret;
```
int ret; on separate line?

ujfalusi replied:
I prefer uninitialized variables in one line.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2290,6 +2294,99 @@ static struct snd_soc_tplg_ops sof_tplg_ops = {
+
+	/* Set nonatomic property for FE dai links as their trigger action involves IPC's */
+	if (!link->no_pcm)
+		link->nonatomic = true;
```
is this needed? The comment above mentions IPCs which are not going to happen with DSP less mode.

But at the same time some BEs such as SoundWire are non-atomic (some delays due to bank switches), and we can't have a atomic FE driver connected to non-atomic BE. We even flag a big warning in soc-pcm.c

````c
 	if (!fe_substream->pcm->nonatomic && be_substream->pcm->nonatomic) {
		dev_err(be->dev, "%s: FE is atomic but BE is nonatomic, invalid configuration\n",
			__func__);
		return -EINVAL;
	}
````

so maybe only update the comment to mention this is not IPC related but to keep the FE-BE compatibility?

ujfalusi replied:
Yes, I will update the comment. We need the nonatomic set for FEs (tested as well).

plbossart commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -111,7 +111,13 @@ int hda_dsp_pcm_hw_params(struct snd_sof_dev *sdev,
-	hstream->format_val = rate | bits | (params_channels(params) - 1);
+	/*
+	 * Use the codec required format val (which is link_bps adjusted) when
+	 * the DSP is not in use
```
the comment is misleading or incorrect. If the DSP is not in use, then the format_val is selected without taking the codec link_pbs into account.

ujfalusi replied:
I think it is the other way around: if the DSP is not in use we take the codec link_bps instead of the bps between the host and the dummy codec.

plbossart reviewed (commented):
minor cosmetic comments below. Thanks @ujfalusi

ujfalusi commented:
Changes since v9:
- patch for sof/pm.c dropped
- patches to allow DSPless mode on all Intel platforms with HDA support added
- Commit message updates
- comment update for the nonatomic flag revised
- rebased on topic/sof-dev (obviously)

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -310,7 +310,7 @@ sof_unprepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widg
 
 	/* skip if the widget is in use or if it is already unprepared */
-	if (!swidget->prepared || swidget->use_count > 0)
+	if (!swidget || !swidget->prepared || swidget->use_count > 1)
```
the > 1 seems wrong? we need the swidget test but the use count should not change, right?

ujfalusi replied:
Right, this was carried over from before the test was fixed. just a sec.

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -1858,7 +1862,7 @@ static int sof_link_load(struct snd_soc_component *scomp, int index, struct snd_
 	}
 
-	/* Set nonatomic property for FE dai links as their trigger action involves IPC's */
+	/* Set nonatomic property for FE dai links for FE-BE compatibility */
```
did you update the comment in the wrong place?

ujfalusi replied:
Yes, I did :(

ranj063 commented on sound/soc/sof/intel/hda.c:
```diff
@@ -865,6 +897,10 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 
 	hda_bus_ml_get_capabilities(bus);
 
+	/* Skip SoundWire if is not supported */
```
typo if "it" is not supported or just if not supported

ujfalusi replied:
OK, thanks

ujfalusi commented:
Changes since v10:
- Update the correct comment  about the nonatomic flag
- Fix the typo in the comment for skipping sndw if it is not available on the platform.

ranj063 reviewed (approved):
looks great. thanks @ujfalusi

plbossart reviewed (changes requested):
only one open with refcounts, looks good otherwise. Thanks @ujfalusi

ujfalusi commented:
Changes since v11:
- correct the use_count test which was carried over from an old revision (in patch 2).

plbossart reviewed (approved):
Thanks @ujfalusi, nice work!
